crates/task-impls/src/upgrade.rs

```diff
@@ -211,35 +231,40 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> UpgradeTaskState<TYPES, I> {
                     }
                 }
             }
-            HotShotEvent::ViewChange(view) => {
-                let view = *view;
-                if *self.cur_view >= *view {
+            HotShotEvent::ViewChange(new_view) => {
+                if self.cur_view >= *new_view {
                     return None;
                 }
 
-                if *view - *self.cur_view > 1 {
-                    warn!("View changed by more than 1 going to view {:?}", view);
-                }
-                self.cur_view = view;
+                self.cur_view = *new_view;
+
+                let view: u64 = *self.cur_view;
+                let time = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .ok()?
+                    .as_secs();
+
                 // We try to form a certificate 5 views before we're leader.
-                if *view >= self.start_proposing_view
-                    && *view < self.stop_proposing_view
-                    && self.quorum_membership.leader(view + 5) == self.public_key
+                if view >= self.start_proposing_view
+                    && view < self.stop_proposing_view
+                    && time >= self.start_proposing_time
+                    && time < self.stop_proposing_time
+                    && self.quorum_membership.leader(TYPES::Time::new(view + 5)) == self.public_key
```

Comment:
Is there any chance this 5, 15, 20, etc could perhaps be constants? I know you have comments, but it looks like the +5 is used in a couple spots where the comment has to be hunted for. It can even just be at the top of this file. Perhaps something like `FORM_CERTIFICATE_OFFSET` or `UPGRADE_OFFSET` or `UPGRADE_INTERVAL`, and commensurately so for the +15 and +20 as something like `UPGRADE_BEGIN_OFFSET` and `UPGRADE_END_OFFSET` and `UPGRADE_DECIDE_BY_OFFSET`.